### DISCUSSION TITLE
Text compression doesn't work

### DISCUSSION THREAD

**Discussion by s1n7ax** [tower-rs/tower-http#489] [Q&A]
### Summary

I'm getting Text Compression warnings in Chrome Lighthouse in a Leptos + Axum SSR project.

![image](https://github.com/tower-rs/tower-http/assets/18459807/c8510dcf-c461-40fe-84f5-b4ff8817de10)

According to [Leptos author](https://github.com/leptos-rs/leptos/discussions/2591#discussioncomment-9419161) tower-http offers compression. I have tried to add it to the best of my knowledge but it seems I haven't done it correctly.

I have following code in Leptos main

```rust
	let compression_layer: CompressionLayer = CompressionLayer::new()
		.gzip(true)
		.deflate(true)
		.br(true)
		.compress_when(DefaultPredicate::new());

	// build our application with a route
	let app = Router::new()
		.image_cache_route(&state)
		.leptos_routes_with_context(&state, routes, state.optimizer.provide_context(), App)
		.layer(compression_layer)
		.fallback(file_and_error_handler)
		.with_state(state);

	let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();
	logging::log!("listening on http://{}", &addr);
	axum::serve(listener, app.into_make_service())
		.await
		.unwrap();
```

After adding the layer, I'm getting additional `content-encoding: gzip` header in the response of the HTML served when `http://localhost:3000/` accessed. However, it's not working for `js` & `css`. Activating the layers does not seem to change the response payload size either.

### Without compression layer

```shell
❯ http GET http://localhost:3000/pkg/my-website.js 'Accept-Encoding: gzip, deflate, br;' -h
HTTP/1.1 200 OK
accept-ranges: bytes
content-length: 52729
content-type: application/javascript
date: Tue, 14 May 2024 16:03:14 GMT
last-modified: Tue, 14 May 2024 16:03:07 GMT

❯ http GET http://localhost:3000/pkg/my-website.js 'Accept-Encoding: gzip, deflate, br;' -b > out

❯ ls out
.rw-r--r-- 53k s1n7ax s1n7ax 14 May 21:33 out
```

### With compression layer

```shell
❯ http GET http://localhost:3000/pkg/my-website.js 'Accept-Encoding: gzip, deflate, br;' -h
HTTP/1.1 200 OK
accept-ranges: bytes
content-length: 52729
content-type: application/javascript
date: Tue, 14 May 2024 16:05:18 GMT
last-modified: Tue, 14 May 2024 16:05:14 GMT

❯ http GET http://localhost:3000/pkg/my-website.js 'Accept-Encoding: gzip, deflate, br;' -b > out

❯ ls out
.rw-r--r-- 53k s1n7ax s1n7ax 14 May 21:35 out
```

Am I doing anything wrong?

### tower-http version

0.2.0

**Comment by jplatte** [comment]
I'm guessing `file_and_error_handler` is the axum handler that serves these js and css files? In that case, just apply the compression layer after registering the fallback handler, otherwise it's not applied for that handler.
